ocaml/xapi/helpers.ml

```diff
@@ -2119,3 +2119,88 @@ let get_active_uefi_certificates ~__context ~self =
       custom_uefi_certs
 
 let uefi_mode_to_string = function `setup -> "setup" | `user -> "user"
+
+module BoundedPsq = struct
+  module type Ordered = sig
+    type t
+
+    val compare : t -> t -> int
+  end
+
+  module type S = sig
+    type t
+
+    type k
+
+    type v
+
+    val create : capacity:int -> t
+
+    val add : t -> k -> v -> unit
+    (** [add t k v] adds mapping [k] => [v] to the priority search queue. If
+        the addition of this mapping would exceed the capacity of the queue,
+        the highest priority (lowest) entry is removed to make space
+        for the newest entry.
+
+        If an entry for [k] is already present, the extant entry is
+        updated with [v]. *)
+
+    val remove : t -> k -> unit
+
+    val clear : t -> unit
+
+    val min : t -> (k * v) option
+
+    val find_opt : k -> t -> v option
+
+    val contains : t -> k -> bool
+
+    val iter : (k -> v -> unit) -> t -> unit
+
+    val size : t -> int
+  end
+
+  module Make (K : Ordered) (V : Ordered) :
+    S with type k = K.t and type v = V.t = struct
+    module Q = Psq.Make (K) (V)
+
+    type k = Q.k
+
+    type v = Q.p
+
+    type t = {capacity: int; mutable queue: Q.t}
+
+    let create ~capacity = {capacity; queue= Q.empty}
+
+    let remove_min ({queue; _} as t) =
+      match Q.min queue with
+      | Some (k, _) ->
+          t.queue <- Q.remove k queue
+      | _ ->
+          ()
+
+    let add t k v =
+      if t.capacity <> 0 then (
```

Comment:
Capacity is allowed to go below zero, the bound here could be changed to:
```suggestion
      if t.capacity < 1 then (
```

Instead, I suggest bounding the capacity on creation, by forcing it to be at least 0:

```ocaml
    let create ~capacity =
      let capacity = Int.max 0 capacity in
      {capacity; queue= Q.empty}
```